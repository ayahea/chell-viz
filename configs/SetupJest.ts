const jestGlobal = global as any;

// tslint:disable-next-line:no-var-requires
jestGlobal.fetch = require('jest-fetch-mock');

// Mock out createObjectURL since it is not available in a fake browser.
jestGlobal.window.URL.createObjectURL = jest.fn();
