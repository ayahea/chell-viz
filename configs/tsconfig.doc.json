{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "declaration": true,
    "outDir": "../docs/api",
    "sourceMap": false
  },
  "include": ["../src/**/*", "../types"],
  "exclude": ["../node_modules", "../**/__tests__/**", "../**/__mocks__/**", "../**/configs/**"],
  "typedocOptions": {
    "mode": "file",
    "out": "docs/api"
  }
}
