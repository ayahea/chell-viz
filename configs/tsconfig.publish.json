{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "declaration": true,
    "outDir": "../lib",
    "sourceMap": false
  },
  "files": ["../chell-viz.ts"],
  "include": ["../src/**/*", "../types", "types"],
  "exclude": ["../node_modules", "../**/__tests__/**", "../**/__mocks__/**", "../configs"]
}
