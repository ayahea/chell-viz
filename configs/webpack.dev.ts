import * as webpack from 'webpack';
import * as merge from 'webpack-merge';
import * as common from '../webpack.common';

module.exports = merge(common, {
  devServer: {
    contentBase: './dist',
    hot: true,
  },
  devtool: 'inline-source-map',
  mode: 'development',
  plugins: [new webpack.HotModuleReplacementPlugin()],
});
