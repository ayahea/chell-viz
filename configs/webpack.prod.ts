import * as UglifyJSPlugin from 'uglifyjs-webpack-plugin';
import * as webpack from 'webpack';
import * as merge from 'webpack-merge';
import * as common from '../webpack.common';

module.exports = merge(common, {
  mode: 'production',
  optimization: {
    splitChunks: {
      chunks: 'all',
    },
  },
  plugins: [
    new UglifyJSPlugin({
      sourceMap: false,
    }),
    new webpack.DefinePlugin({
      'process.env.NODE_ENV': JSON.stringify('production'),
    }),
  ],
});
