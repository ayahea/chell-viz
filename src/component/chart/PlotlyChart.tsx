import * as Immutable from 'immutable';
import * as React from 'react';

// We need to use dist/plotly to avoid forcing users to do some webpack gymnastics:
// https://github.com/plotly/plotly-webpack#the-easy-way-recommended
import * as plotly from 'plotly.js/dist/plotly';

import { withDefaultProps } from '../../helper/ReactHelper';

export enum PLOTLY_CHART_TYPE {
  /** [Plotly Bar Chart](https://plot.ly/javascript/bar-charts/) */
  'bar' = 'bar',
  /** [Plotly Point Cloud](https://plot.ly/javascript/pointcloud/) */
  'pointcloud' = 'pointcloud',
  /** [Plotly Line/Scatter Chart](https://plot.ly/javascript/line-and-scatter/) */
  'scatter' = 'scatter',
  /** [Plotly Line/Scatter Chart in WebGL](https://plot.ly/javascript/line-and-scatter/) */
  'scattergl' = 'scattergl',
  /** [Plotly 3D Scatter Plot](https://plot.ly/javascript/3d-scatter-plots/) */
  'scatter3d' = 'scatter3d',
}

export interface IPlotlyData extends plotly.ScatterData {
  type: PLOTLY_CHART_TYPE | 'bar' | 'pointcloud' | 'scatter' | 'scattergl' | 'scatter3d';
}

export const defaultPlotlyConfig: Partial<Plotly.Config> = {
  displayModeBar: false,
  scrollZoom: true,
  // modeBarButtons: [['zoomOut2d', 'zoomIn2d'], ['resetScale2d', 'autoScale2d'], ['select2d', 'pan2d']],
};

export const defaultPlotlyLayout: Partial<Plotly.Layout> = {
  autosize: true,
  height: 400,
  hovermode: 'closest',
  legend: {},
  margin: {
    b: 10,
    l: 40,
    r: 10,
    t: 10,
  },
  showlegend: false,
  title: '',
  width: 400,
  xaxis: {
    range: [30],
  },
  yaxis: {
    range: [30],
  },
};

export const defaultPlotlyChartProps = {
  config: {} as Partial<Plotly.Config>,
  data: [] as Array<Partial<IPlotlyData>>,
  layout: {} as Partial<Plotly.Layout>,
  onClickCallback: (event: plotly.PlotMouseEvent) => {
    return;
  },
  onHoverCallback: (event: plotly.PlotMouseEvent) => {
    return;
  },
  onSelectedCallback: (event: plotly.PlotSelectionEvent) => {
    return;
  },
  onUnHoverCallback: (event: plotly.PlotMouseEvent) => {
    return;
  },
};

export type PlotlyChartProps = {
  data: Array<Partial<IPlotlyData>>;
} & Partial<typeof defaultPlotlyChartProps>;

/**
 * React wrapper for a Plotly Chart.
 *
 * @description
 * Based upon: https://github.com/davidctj/react-plotlyjs-ts
 *
 * @export
 * @extends {React.Component<IPlotlyChartProps, any>}
 */
export class PlotlyChartClass extends React.Component<PlotlyChartProps, any> {
  public plotlyCanvas: plotly.PlotlyHTMLElement | null = null;
  protected canvasRef: HTMLDivElement | null = null;

  public attachListeners() {
    this.plotlyCanvas!.on('plotly_click', this.onClick);
    this.plotlyCanvas!.on('plotly_selected', this.onSelect);

    this.plotlyCanvas!.on('plotly_hover', this.onHover);
    this.plotlyCanvas!.on('plotly_unhover', this.onUnHover);

    window.removeEventListener('resize', this.resize);
    window.addEventListener('resize', this.resize);
  }

  public resize = () => {
    plotly.Plots.resize(this.plotlyCanvas!);
  };

  public draw = async () => {
    const { data, layout, config } = this.props;
    if (this.plotlyCanvas) {
      // plotly.react will not destroy the old plot: https://plot.ly/javascript/plotlyjs-function-reference/#plotlyreact
      this.plotlyCanvas = await plotly.react(
        this.plotlyCanvas,
        data,
        this.getMergedLayout(layout),
        this.getMergedConfig(config),
      );
    }
  };

  public componentDidUpdate(prevProps: PlotlyChartProps) {
    const { data, layout, config } = this.props;
    if (data !== prevProps.data || layout !== prevProps.layout || config !== prevProps.config) {
      this.draw();
    }
  }

  public async componentDidMount() {
    if (this.canvasRef && !this.plotlyCanvas) {
      const { data, layout, config } = this.props;
      this.plotlyCanvas = await plotly.react(
        this.canvasRef,
        data,
        this.getMergedLayout(layout),
        this.getMergedConfig(config),
      );
      this.attachListeners();
      this.draw();
    }
  }

  public componentWillUnmount() {
    if (this.plotlyCanvas) {
      plotly.purge(this.plotlyCanvas);
      this.plotlyCanvas = null;
    }
    window.removeEventListener('resize', this.resize);
  }

  public render() {
    return (
      <div className={'plotly-chart'} ref={node => (this.canvasRef = node ? node : null)} style={{ marginBottom: 5 }} />
    );
  }

  protected getMergedConfig = (config: Partial<Plotly.Config> = {}) => {
    const foo: Immutable.Map<any, any> = Immutable.fromJS(Object.assign({}, defaultPlotlyConfig));
    const bar: Immutable.Map<any, any> = Immutable.fromJS(Object.assign({}, config));
    return Object.assign({}, foo.mergeDeep(bar).toJS());
  };

  protected getMergedLayout = (layout: Partial<Plotly.Layout> = {}) => {
    const foo: Immutable.Map<any, any> = Immutable.fromJS(Object.assign({}, defaultPlotlyLayout));
    const bar: Immutable.Map<any, any> = Immutable.fromJS(Object.assign({}, layout));
    return Object.assign({}, foo.mergeDeep(bar).toJS());
  };

  protected onClick = (event: plotly.PlotMouseEvent) => {
    const { onClickCallback } = this.props;
    if (onClickCallback) {
      onClickCallback(event);
    }
  };

  protected onHover = (event: plotly.PlotMouseEvent) => {
    const { onHoverCallback } = this.props;
    if (onHoverCallback) {
      onHoverCallback(event);
    }
  };

  protected onSelect = (event: plotly.PlotSelectionEvent) => {
    const { onSelectedCallback } = this.props;
    if (onSelectedCallback) {
      onSelectedCallback(event);
    }
  };

  protected onUnHover = (event: plotly.PlotMouseEvent) => {
    const { onUnHoverCallback } = this.props;
    if (onUnHoverCallback) {
      onUnHoverCallback(event);
    }
  };
}

const PlotlyChart = withDefaultProps(defaultPlotlyChartProps, PlotlyChartClass);

export default PlotlyChart;
export { PlotlyChart };
