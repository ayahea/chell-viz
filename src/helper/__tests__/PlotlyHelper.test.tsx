import { CommonWrapper, mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as React from 'react';

import { IMockDict } from 'configs/SetupJest';
import PlotlyChart, { generatePointCloudData, IPlotlyChartProps } from '../../helper/PlotlyHelper';

beforeEach(() => {
  jest.resetModules();
});

describe('PlotlyChart', () => {
  const sampleData = [
    {
      marker: {
        color: 'blue',
      },
      mode: 'markers',
      type: 'pointcloud',
      uid: 'adam',
      xy: new Float32Array([1, 2, 3, 4]),
    } as Partial<Plotly.ScatterData>,
  ];

  /**
   * Helper function to create and wait for a PlotlyChart to be mounted.
   *
   * @param props Custom props to be passed to the chart.
   * @returns A wrapper for the PlotlyChart that has been mounted.
   */
  const getMountedPlotlyChart = async (props: IPlotlyChartProps) => {
    const wrapper = mount(<PlotlyChart {...props} />);
    await wrapper.update();
    return wrapper;
  };

  /**
   * Helper function to dispatch an event through plotly.
   *
   * @param wrapper The PlotlyChart.
   * @param event The name of the event to dispatch.
   */
  const dispatchPlotlyEvent = (wrapper: CommonWrapper, event: string) => {
    (wrapper.instance() as PlotlyChart).plotlyCanvas!.dispatchEvent(new Event(event));
  };

  test('Should match existing snapshot when given empty data.', () => {
    const wrapper = mount(<PlotlyChart data={[]} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should match existing snapshot when given sample data.', () => {
    const wrapper = mount(<PlotlyChart data={sampleData} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should handle callbacks.', () => {
    const spies: IMockDict = {
      onClickSpy: jest.fn(),
      onHoverSpy: jest.fn(),
      onSelectedSpy: jest.fn(),
      onUnHoverSpy: jest.fn(),
    };
    const wrapper = shallow(
      <PlotlyChart
        data={sampleData}
        onClickCallback={spies.onClickSpy}
        onHoverCallback={spies.onHoverSpy}
        onSelectedCallback={spies.onSelectedSpy}
        onUnHoverCallback={spies.onUnHoverSpy}
      />,
    );

    wrapper.instance().props.onClickCallback();
    wrapper.instance().props.onHoverCallback();
    wrapper.instance().props.onSelectedCallback();
    wrapper.instance().props.onUnHoverCallback();

    for (const key of Object.keys(spies)) {
      expect(spies[key]).toHaveBeenCalledTimes(1);
    }
  });

  test('Should attach events when the container is attached.', async () => {
    const spies: IMockDict = {
      onClickSpy: jest.fn(),
      onSelectedSpy: jest.fn(),
    };

    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
      onClickCallback: spies.onClickSpy,
      onSelectedCallback: spies.onSelectedSpy,
    });

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should call the appropriate callback when the window is resized.', async () => {
    const onResizeSpy = jest.fn();

    const wrapper = await getMountedPlotlyChart({ data: sampleData });
    const chartInstance = wrapper.instance() as PlotlyChart;

    chartInstance.resize = onResizeSpy;
    chartInstance.attachListeners();
    window.dispatchEvent(new Event('resize'));
    expect(onResizeSpy).toHaveBeenCalledTimes(1);
  });

  test('Should call the appropriate callback when plotly emits a click event.', async () => {
    const onClickSpy = jest.fn();
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
      onClickCallback: onClickSpy,
    });
    dispatchPlotlyEvent(wrapper, 'plotly_click');
    expect(onClickSpy).toBeCalled();
  });

  test('Should call the appropriate callback when plotly emits a hover event.', async () => {
    const onHoverSpy = jest.fn();
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
      onHoverCallback: onHoverSpy,
    });
    dispatchPlotlyEvent(wrapper, 'plotly_hover');
    expect(onHoverSpy).toBeCalled();
  });

  test('Should call the appropriate callback when plotly emits a un-hover event.', async () => {
    const onUnHoverSpy = jest.fn();
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
      onUnHoverCallback: onUnHoverSpy,
    });
    dispatchPlotlyEvent(wrapper, 'plotly_unhover');
    expect(onUnHoverSpy).toBeCalled();
  });

  test('Should call the appropriate callback when plotly emits a selected event.', async () => {
    const onSelectedSpy = jest.fn();
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
      onSelectedCallback: onSelectedSpy,
    });
    dispatchPlotlyEvent(wrapper, 'plotly_selected');
    expect(onSelectedSpy).toBeCalled();
  });

  test('Should unmount correctly.', async () => {
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
    });

    const chartInstance = wrapper.instance() as PlotlyChart;

    expect(chartInstance.plotlyCanvas).not.toBeNull();
    wrapper.unmount();
    expect(chartInstance.plotlyCanvas).toBeNull();
  });
});

describe('generatePointCloudData', () => {
  test('Should create the expected plotly data format when given defaults.', () => {
    const result = generatePointCloudData(new Float32Array([1, 2, 3]), 'red', 4);
    expect(result).toMatchSnapshot();
  });
});
