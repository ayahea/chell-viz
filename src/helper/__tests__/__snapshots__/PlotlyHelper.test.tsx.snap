// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlotlyChart Should attach events when the container is attached. 1`] = `
<PlotlyChart
  data={
    Array [
      Object {
        "marker": Object {
          "color": "blue",
        },
        "mode": "markers",
        "type": "pointcloud",
        "uid": "adam",
        "xy": Float32Array [
          1,
          2,
          3,
          4,
        ],
      },
    ]
  }
  onClickCallback={[MockFunction]}
  onSelectedCallback={[MockFunction]}
>
  <div />
</PlotlyChart>
`;

exports[`PlotlyChart Should match existing snapshot when given empty data. 1`] = `
<PlotlyChart
  data={Array []}
>
  <div />
</PlotlyChart>
`;

exports[`PlotlyChart Should match existing snapshot when given sample data. 1`] = `
<PlotlyChart
  data={
    Array [
      Object {
        "marker": Object {
          "color": "blue",
        },
        "mode": "markers",
        "type": "pointcloud",
        "uid": "adam",
        "xy": Float32Array [
          1,
          2,
          3,
          4,
        ],
      },
    ]
  }
>
  <div />
</PlotlyChart>
`;

exports[`generatePointCloudData Should create the expected plotly data format when given defaults. 1`] = `
Object {
  "marker": Object {
    "color": "red",
    "sizemax": 8,
    "sizemin": 4,
  },
  "mode": "markers",
  "type": "pointcloud",
  "xy": Float32Array [
    1,
    2,
    3,
    3,
    2,
    1,
  ],
}
`;
